### PR TITLE
feat #372 & #373

### LINKED ISSUE #372: `TimeTranslationPastHorizon` exception for private testnet
## Example failure 

https://github.com/geniusyield/atlas/actions/runs/12329517611/job/34413921377?pr=370.

## Description

Sometimes we may encounter `TimeTranslationPastHorizon` exception when running a test against private testnet. Though the functions we expose inside Atlas to do slot to time conversations successfully give result even if provided slot is beyond the era end, we might still see this exception when slot to time conversion is occuring when Plutus script context is being built inside `cardano-api` machinery. For example, if in our off-chain code, we say that transaction must be invalid after a particular slot then when this slot is converted to posix time inside `cardano-api` library, it can throw exception if this slot is beyond the era end as ultimately `slotToWallclock` from `ouroboros-consensus` library is called (function trace is provided below) which doesn't allow such conversations as era parameters could change beyond the reported era end. Note that reported era end is that epoch until we know that no such updates can occur.

```
makeTransactionBodyAutoBalance -> evaluateTransactionExecutionUnits -> evaluateTransactionExecutionUnitsShelley -> evalTxExUnitsWithLogs -> mkPlutusWithContext -> .. -> .. -> toPlutusTxInfo -> transValidityInterval -> transSlotToPOSIXTime -> slotToPOSIXTime -> epochInfoSlotToUTCTime ~ slotToWallclock (via interpreterToEpochInfo in ouroboros-consensus)
```

Likely we can modify era end to be unbounded for last era?

### LINKED ISSUE #373: allow `addRefScript` to work with higher plutus versions

### PR DESCRIPTION
**Pull request by sourabhxyz**
Closes #372, closes #373.

Now era histories are cached through entire run of the program whereas protocol parameters are cached per epoch.

Inside `mkMethod` of `makeGetParameters` I had added print statements [here](https://github.com/geniusyield/atlas/blob/6d78550f149407cd19aad8b8291c91ce286f66f9/src/GeniusYield/Types/Providers.hs#L423-L431) to track values of the associated `MVar` to verify update is correctly occurring, see below logs:

```
Serving cached parameters, currentTime: 2024-12-18 10:02:46.504257 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:02:55.519935 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:04.54132 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:07.552586 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:10.560332 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:16.579995 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:19.596804 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:22.631832 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:25.660371 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC
Refreshing parameters, currentTime: 2024-12-18 10:03:31.672819 UTC and nextEpochBeginTime: 2024-12-18 10:03:28 UTC and timeDelta: 50s
newNextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:37.714951 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:46.755501 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:52.773923 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:03:58.797953 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:04:01.817292 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
Serving cached parameters, currentTime: 2024-12-18 10:04:04.829536 UTC and nextEpochBeginTime: 2024-12-18 10:04:18 UTC
```

This however is removed from final PR as we don't want to perform such IO action inside `MVar` update.

I had also written a privnet test file to verify that new fetch is occurring in next epoch [here](https://github.com/geniusyield/atlas/blob/6d78550f149407cd19aad8b8291c91ce286f66f9/tests-privnet/GeniusYield/Test/Privnet/Epoch.hs) however this is removed from final PR as it's not automated.

While working on this task, I had noticed a nasty bug where our caching mechanism was essentially broken since when we call methods like [`getProtParams'`](https://github.com/geniusyield/atlas/blob/5dd92790cbedb5604c474b319703b8cb5ca0a266/src/GeniusYield/Types/Providers.hs#L411) we are actually generating fresh `MVar` instead of utilising an old one.

### CODE REVIEW

**Review by 4TT1L4 — DISMISSED** [review]
LGTM

**Review by 4TT1L4 — DISMISSED** [review]
Thanks for fixing the CI build

Re-approved.

File: src/GeniusYield/Providers/Common.hs [hunk 1]
```diff
@@ -119,6 +121,20 @@ fromJson b = do
...
+makeLastEraEndUnbounded eh =
+  let Ouroboros.Summary eraList = extractEraSummaries eh
+   in Api.EraHistory $ Ouroboros.mkInterpreter $ Ouroboros.Summary $ g eraList
+ where
+  g eraList =
+    let eraList' = nonEmptyToList eraList
+        f [] = []
+        f [x] =
+          let oldEraParams = Ouroboros.eraParams x
+           in [x {Ouroboros.eraEnd = Ouroboros.EraUnbounded, Ouroboros.eraParams = oldEraParams {Ouroboros.eraSafeZone = Ouroboros.UnsafeIndefiniteSafeZone}}]
```
Comment (sourabhxyz) [inline src/GeniusYield/Providers/Common.hs:134]: Utilising `UnsafeIndefiniteSafeZone` is perhaps not required as when I do `IOG.epochInfoSlotToRelativeTime_ ei (slotToApi $ slotFromWord64 5000000)` without it inside privnet, I am able to get successful result.

File: src/GeniusYield/Test/Privnet/Setup.hs [hunk 2]
```diff
@@ -76,7 +76,7 @@ The first argument is the log severity filter. Only logs of this severity or hig
 newtype Setup = Setup (GYLogSeverity -> (String -> IO ()) -> (Ctx -> IO ()) -> IO ())
 
 cardanoDefaultTestnetOptionsConway :: CardanoTestnetOptions
-cardanoDefaultTestnetOptionsConway = cardanoDefaultTestnetOptions {cardanoNodeEra = Api.AnyCardanoEra Api.ConwayEra}
+cardanoDefaultTestnetOptionsConway = cardanoDefaultTestnetOptions {cardanoNodeEra = Api.AnyCardanoEra Api.ConwayEra, cardanoEpochLength = 2000}
```
Comment (sourabhxyz) [inline src/GeniusYield/Test/Privnet/Setup.hs:79]: As noted in [this](https://github.com/geniusyield/atlas/issues/372#issuecomment-2551249498) comment, I think it's best to just increase epoch length.

**Review by 4TT1L4 — APPROVED** [review]
LGTM

**Pull request merged**